### PR TITLE
Environment Info

### PR DESCRIPTION
**Pull request by rgomezp**
### `EnvironmentInfoHelper`
- new helper encapsulates all Runtime Environment Information into single source
- `EnvironmentInfo` includes 
   - `isHttps` 
   - `browser`
   - `version  `
   - `shouldAutoAccept`
- logic for `autoAccept` also lives in helper class private function. In the future, logic for environment-specific parameters can be added as helper class functions

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-website-sdk/585)

### CODE REVIEW

**Comment by rgomezp** [comment]
### `!!! CHANGE BASE TO MASTER BEFORE MERGING !!!`

File: src/helpers/EnvironmentInfoHelper.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        // from bowser source: https://bit.ly/36mq1R5
+        return (BROWSER_TYPES as any)[bowser.name];
+    }
+
+    private static getBrowserVersion(): number {
+        return Number(bowser.version);
+    }
+
+    private static isHttps(): boolean {
+        return location.protocol == 'https:';
```
Comment (jkasten2) [inline src/helpers/EnvironmentInfoHelper.ts]: nit, should be `===`.

File: src/helpers/EnvironmentInfoHelper.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private static isUsingSubscriptionWorkaround(): boolean {
+        return OneSignalUtils.isUsingSubscriptionWorkaround();
+    }
+    
+    private static isBrowserAndSupportsServiceWorkers(): boolean {
+        return (this.isBrowser() && navigator && 'serviceWorker' in navigator);
```
[the remark this answers is not in the document]
Comment (jkasten2) [reply src/helpers/EnvironmentInfoHelper.ts]: A whitelist of browsers seems fragile, we should just keep doing feature detection. Or a black list if it is really needed.

File: src/helpers/EnvironmentInfoHelper.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        // Firefox 72+ requires user-interaction. For HTTP prompt to work,
+        // we need to set autoAccept to false
+        if (this.getBrowser() === "firefox" && this.getBrowserVersion() < 72) {
+            requiresUserInteraction = false;
+        }
+        return requiresUserInteraction;
+    }
+
+    private static getOsVersion(): number {
+        return Number(bowser.osversion);
```
Comment (jkasten2) [inline src/helpers/EnvironmentInfoHelper.ts]: `bowser.osversion` is defined as `string|number` Number might not be safe.

File: src/helpers/EnvironmentInfoHelper.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    private static isUsingSubscriptionWorkaround(): boolean {
+        return OneSignalUtils.isUsingSubscriptionWorkaround();
+    }
+    
+    private static isBrowserAndSupportsServiceWorkers(): boolean {
+        return (this.isBrowser() && navigator && 'serviceWorker' in navigator);
+    }
+
+    private static requiresUserInteraction(): boolean {
+        var requiresUserInteraction = true;
```
Comment (jkasten2) [inline src/helpers/EnvironmentInfoHelper.ts]: Default should be `false` by default. Only Safari 12.1+ and Firefox 72+ require user interaction to prompt.

Comment (jkasten2) [reply src/helpers/EnvironmentInfoHelper.ts]: We could also clean up some of the checks we have for this now in production code.

File: express_webpack/README.md [hunk 5]
```diff
@@ -70,9 +70,4 @@ This sets the BUILD environment origin to `texas` which will result in SDK files
    - HTTP: `4000`
    - HTTPS: `4001`
 
-#### NOTE ON PORTS:
-**SDK**: SDK files will automatically be fetched from the 4000s ports depending on the HTTP/S setting
-   - HTTP: `4000`
-   - HTTPS: `4001`
-
-**API**: dev-environment API calls will be made to the `3001` port (e.g: `<custom-origin>:3001`)
+**API**: dev-environment API calls will be made to the `3001` port (e.g: `<custom-origin>:3001`)
```
[the remark this answers is not in the document]
Comment (rgomezp) [reply express_webpack/README.md]: This is a duplicate, so its fine to remove here

File: src/OneSignal.ts [hunk 6]
```diff
@@ -781,6 +785,7 @@ export default class OneSignal {
   static VERSION = Environment.version();
   static _VERSION = Environment.version();
   static sdkEnvironment = SdkEnvironment;
+  static environmentInfo ?: EnvironmentInfo = undefined;
```
Comment (jkasten2) [inline src/OneSignal.ts]: Can leave off `undefined`.
Remove space before ?

**Pull request merged**